src/matrixrtc/MatrixRTCSession.ts

```diff
@@ -742,46 +826,54 @@ export class MatrixRTCSession extends TypedEventEmitter<MatrixRTCSessionEvent, M
         if (!localUserId || !localDeviceId) throw new Error("User ID or device ID was null!");
 
         const myCallMemberEvent = roomState.getStateEvents(EventType.GroupCallMemberPrefix, localUserId) ?? undefined;
-        const content = myCallMemberEvent?.getContent<Record<any, unknown>>() ?? {};
-        const memberships: CallMembershipData[] = Array.isArray(content["memberships"]) ? content["memberships"] : [];
-
-        const myPrevMembershipData = memberships.find((m) => m.device_id === localDeviceId);
-        let myPrevMembership;
-        try {
-            if (myCallMemberEvent && myPrevMembershipData && myPrevMembershipData.membershipID === this.membershipId) {
-                myPrevMembership = new CallMembership(myCallMemberEvent, myPrevMembershipData);
+        const content = myCallMemberEvent?.getContent<Record<string, unknown>>() ?? {};
+        const legacy = "memberships" in content || this.useLegacyMemberEvents;
+        let newContent: {} | ExperimentalGroupCallRoomMemberState | SessionMembershipData = {};
+        if (legacy) {
+            let myPrevMembership;
+            // We know its CallMembershipDataLegacy
+            const memberships = Array.isArray(content["memberships"])
+                ? content["memberships"]
+                : ([] as CallMembershipDataLegacy[]);
+            const myPrevMembershipData = memberships.find((m) => m.device_id === localDeviceId);
+            try {
+                if (
+                    myCallMemberEvent &&
+                    myPrevMembershipData &&
+                    isLegacyCallMembershipData(myPrevMembershipData) &&
+                    myPrevMembershipData.membershipID === this.membershipId
+                ) {
+                    myPrevMembership = new CallMembership(myCallMemberEvent, myPrevMembershipData);
+                }
+            } catch (e) {
+                // This would indicate a bug or something weird if our own call membership
+                // wasn't valid
+                logger.warn("Our previous call membership was invalid - this shouldn't happen.", e);
             }
```

Comment:
Since what throws here is the `CallMembership` constructor, it would help to focus the try-catch block.  The constructor also has its own call to `isLegacyCallMembershipData`, so it can be removed from here.
```suggestion
            if (
                myCallMemberEvent &&
                myPrevMembershipData &&
                myPrevMembershipData.membershipID === this.membershipId
            ) {
                try {
                    myPrevMembership = new CallMembership(myCallMemberEvent, myPrevMembershipData);
                } catch (e) {
                    // This would indicate a bug or something weird if our own call membership
                    // wasn't valid
                    logger.warn("Our previous call membership was invalid - this shouldn't happen.", e);
                }
            }
```

Reply:
It would change the logic a little bit however. removing the isLegacyCallMembershipData would allow a valid construction if an event formatted like a session Memberhsip would also work (but should not work).

Reply:
If it's important enough to error-check the membership object, then it's inconsistent to say that we know its `CallMembershipDataLegacy` and assign a type to `memberships`.

Then again, the runtime checks balance out the risks of any "eager" type hints.  So I'm okay with leaving this as it is.